blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineRunImpl.java

```diff
@@ -135,11 +143,33 @@ private boolean isReplayable(ReplayAction replayAction) {
         return replayAction != null && replayAction.isRebuildEnabled();
     }
 
+    static final long PIPELINE_NODE_CONTAINER_CACHE_MAX_SIZE = Long.getLong("PIPELINE_NODE_CONTAINER_CACHE_MAX_SIZE", 10000);
+
+    private static Cache<String, PipelineNodeContainerImpl> PIPELINE_NODE_CONTAINER_LOADING_CACHE = Caffeine.newBuilder()
+        .maximumSize(PIPELINE_NODE_CONTAINER_CACHE_MAX_SIZE)
+        .expireAfterAccess(1, TimeUnit.DAYS)
+        .build();
+
+    /**
+     * for test purpose
+     */
+    static void clearCache() {
+        PIPELINE_NODE_CONTAINER_LOADING_CACHE.invalidateAll();
+    }
+
     @Override
     @Navigable
     public BluePipelineNodeContainer getNodes() {
         if (run != null) {
-            return new PipelineNodeContainerImpl(run, getLink());
+            // not using cache if run not completed
+            if (!run.isBuilding() ) {
```

Comment:
You might want
```suggestion
            // not using cache if run not completed
            if (!run.isLogUpdated() ) {
```
which is subtly different.

Reply:
done https://github.com/jenkinsci/blueocean-plugin/commit/ae3da2701919e5d6c4dfa1fecd054754f247e2d8